### ISSUE TITLE
sclang WebView/HelpBrowser can't evaluate code

### ISSUE THREAD

**Issue by adcxyz** [supercollider/supercollider#4000]
Environment
-----------

*  SuperCollider version: 3.10.beta1
* macOS 10.13.6

Steps to reproduce (for bugs)
-----------------------------

```supercollider
// open a help file
SinOsc.help; 
// run the first example in the browser:
{ SinOsc.ar(200, 0, 0.5) }.play;
// returns 2 synths instead of one!
-> Synth('temp__0' : 1024)
-> Synth('temp__1' : 1025)
```

Strangely enough, WebView does not seem interpret selected text when told to:
```
v = WebView(nil, 800@800).front;
v.url_(Platform.userAppSupportDir +/+ "Help/Classes/SinOsc.html");
v.enterInterpretsSelection = true;
v.front;
// -> evaluating code in the browser window does nothing
```

**Comment by brianlheim** [comment]
Dupe of #3924. Please check open issues before filing!

**Comment by brianlheim** [comment]
The first part of this issue is already covered by #3924, please remove it from this ticket.

For the second part, what do you mean specifically by evaluate? We have two shortcuts for doing that. Did this used to work? (I genuinely don't know)

**Comment by brianlheim** [comment]
Also, sorry for the accidental close; I misread this at first at thought it was only a duplicate. My mistake!

**Comment by adcxyz** [comment]
No problem - I had checked for open issues with help browser, but not help documents, so I did not see #3924.

**Comment by brianlheim** [comment]
OK! This issue still needs to be updated/more info as per comment above though.

**Comment by prko** [comment]
Has 3.10.0-rc1 still the problem?

**Comment by brianlheim** [comment]
yes

**Comment by redFrik** [comment]
just ran into this with 3.10.2-rc1 on macOS.  i select any code in scdoc and it evaluates twice.

**Comment by sebshader** [comment]
this occurs for me as well (3.10.2) on OSX on certain examples. For instance, in the Introduction to GUI guide help file everything is fine until this example:
`
w = Window.new("GUI Introduction").layout_(
    VLayout(
        HLayout( Button(), TextField(), Button() ),
        TextView()
    )
).front;
`
& I see 2 "Window"s in the post window (& get 2 windows)

**Comment by snappizz** [comment]
i think this was fixed recently, can people retry in 3.10 branch?

**Comment by patrickdupuis** [comment]
I've tested on linux and could not reproduce. All of these reports are on OSX/macOS however.

**Comment by snappizz** [comment]
the first half of the issue is a duplicate of #3924, which is fixed.

the second half of the issue appears legitimate.

this issue is too confusing, so i am closing it and opening a new one.

**Comment by snappizz** [comment]
thinking about this more.

the issue is that you can't execute code from a WebView or HelpBrower, only the IDE help browser. fixing this will require adding WebChannel support to sclang, and presents a fairly significant time investment.

i've always felt that the ability to execute code from the help browser is an unnecessary and gratuitous feature. if you need to run code, just copy it to the IDE window. maintaining all these implementations of code execution isn't a good use of our time in my opinion.

others are free to disagree, but this is a wontfix for me.

**Comment by tremblap** [comment]
Hmmmm, labeling it a wontfix, despite being a feature that was present since the begining and is used a lot by most of my students (and myself) is a daring gesture... should we open a discussion on the mailing list to see if it is as used as I think it is (and therefore and essential fix?)

**Comment by brianlheim** [comment]
> should we open a discussion on the mailing list to see if it is as used as I think it is (and therefore and essential fix?)

This sounds very petty and is not likely to get you any enthusiasm from devs. Please don't do this, it is just going to make people miserable.

I do think it's premature to label this as wontfix @snappizz , and I am going to reopen it in the hope that we can solve it in the future. In addition, I've seen a request for adding websocket support to sclang, and adding new Qt class wrappers is relatively straightforward, so it would probably get use beyond this one case. But @tremblap, I hope you will try to be a little more understanding about the concerns project maintainers have to balance. When @snappizz says this would take a significant time investment to solve, he isn't kidding, and there are lots of other issues we could solve in the meanwhile.

**Comment by brianlheim** [comment]
Or -- could we solve this with a custom 'onJavaScripMsg_' handler? Is that too hacky?

**Comment by snappizz** [comment]
my feeling (not a ruling) on this is that supercollider is a computer music language for synthesizing sound and making music. a web browser built into it is already feature creep imo. code execution from an integrated help system that duplicates functionality of the IDE is serious feature creep.

the only reason we have code eval from the help browser is because SC used to use RTF code files for both help and for code documents, so it was easy to implement. with the introduction of SCDoc, it is significantly more complex to implement, and now we're saddled with maintaining two-going-on-three implementations of sclang code eval (IDE, IDE help browser, now sclang help browser).

this is an auxiliary feature whose value to users is sentimental, and has presented a significant burden on development recently (#3877, #3960, #4177, #4256, #4267). i believe maintaining it is a poor use of our time and i would even push for its removal entirely. a button that lets you easily copy the code to the clipboard so you can paste it into the IDE is plenty good, and not an unreasonable request considering that it's what every other documentation system does.

historically, SC was way too eager to add new features because they seemed fun without weighing the possible long-term consequences of maintaining them. help browser code evaluation is to me exemplary of bloat in SC, and letting go of such unnecessary features will benefit long-term project health.

**Comment by jamshark70** [comment]
I'd suggest to be careful about messaging. I tend to agree that it's gotten too complicated, and it would be better not to promise in-browser code execution at all than to promise it when it doesn't work reliably. At the same time, users will perceive this as a degradation in user experience, and if the messaging is along the lines of "you shouldn't want this" (let's not kid ourselves, this comes through plenty strongly between the lines) then it will not fly very well.

One counterexample is that we do get users switching over from Max or Pure Data, where help patches *are* fully interactive. Of course we don't have to mimic that, but there's precedent for interactive help in music environments.

Another argument in favor of removing code-interaction features from help is that it feels very strange to me to have editable code fields in the browser. If you can edit it, you might want to persist it, but the browser doesn't allow that (for good reason)... much clearer to pop the code over to a window that is obviously for code editing (and has full autocompletion etc.).

> a button that lets you easily copy the code to the clipboard so you can paste it into the IDE is plenty good, and not an unreasonable request considering that it's what every other documentation system does.

Or, a step further, a button that opens a "help code" Document and automatically pastes it in for you...? Remove a step for the user. If the user doesn't have to paste it manually, that would help user acceptance IMO.

**Comment by tremblap** [comment]
first, apologies if my message sounded like a threat. 2nd language English here, and it was more meant as a surprise and worried that a most important feature for newbies and even intermediates like me (aka the help browser in the IDE being able to execute the example code) was to be dismissed so easily as in this sentence: 

> if you need to run code, just copy it to the IDE window. 

I am not certain what is what, but this is what I reported originally (and that was closed as a duplicate to this one

**Comment by tremblap** [comment]
Strange coincidence: a certain Daniel Mayer posted on the list yesterday about this! He is sending a lot of love to the dev team, so I thought I'd point it out ;-)

**Comment by jamshark70** [comment]
It kind of seems like the user perspective may be underestimating the dev cost to maintain this feature (and perhaps also overestimating its value) while the dev perspective may be underestimating how well liked the feature is.

@tremblap one useful question is: given that SuperCollider is a REPL-type environment, what do other REPLs do about help code execution? If it's really a critical feature, then a majority of REPL environments will take pains to make it easy to run code from within help documents directly. I'm willing to bet that they don't, and users of those environments get along just fine. That is, perhaps it seems like a critical feature because you're used to it and you like it, but maybe it isn't actually indispensable.

Suppose scdoc browser code execution is removed. Future new users will not know that. I'm willing to bet that a good percentage of them, if you told them "copy the code to run" or even better "click this button and the code will appear instantly in a place where you can run it," would probably accept that.

For devs, I'll reiterate that this is a well liked feature. Be careful how you talk about removing it. I think there's a good case to be made for dropping it but we see here already how sensitive this issue can be.

**Comment by brianlheim** [comment]
@tremblap apology accepted. My criticism was of the course of action you suggested, and not your phrasing of it. My apologies if it read as a criticism of your English proficiency.

@snappizz @jamshark70 i see your points, but i don't think this is a good place to rehash the entire debate about the HelpBrowser. let's please stay on topic and move any larger discussions to a separate ticket or other venue.

**Comment by tremblap** [comment]
> My apologies if it read as a criticism of your English proficiency.

No it was more that I misread the discussion and proposal...

**Comment by brianlheim** [comment]
fixed in #4883, thanks @elgiano !
